### DISCUSSION TITLE
How to use the streaming API remotely

### DISCUSSION THREAD

**Discussion by Mu2u** [oobabooga/text-generation-webui#1605] [Q&A]
Hi is there a way to use the streaming websockets api but use it from a remote server?  I typically use Cloudflare Tunnel for the base API but WS is not supported for cloudflare (at least not fast setup)

Does anyone else have this working and can explain how they have it connected to an app like Silly TavernAI?

**Comment by Daviljoe193** [comment]
I hope you're still looking for an answer, because I did make something work. Basically, use RemoteMoe. In Colab, the codeblock would look something like this.

```
#@title 2.5. API tunnel (Optional)

import os
import shlex
import subprocess
from pathlib import Path
from typing import Union
import threading

!pkill ssh

id_rsa_file = "/content/id_rsa"
id_rsa_pub_file = "/content/id_rsa.pub"
if os.path.exists(id_rsa_file):
    os.remove(id_rsa_file)
if os.path.exists(id_rsa_pub_file):
    os.remove(id_rsa_pub_file)

def gen_key(path: Union[str, Path]) -> None:
    path = Path(path)
    arg_string = f'ssh-keygen -t rsa -b 4096 -N "" -q -f {path.as_posix()}'
    args = shlex.split(arg_string)
    subprocess.run(args, check=True)
    path.chmod(0o600)

ssh_name = "id_rsa"
ssh_path = Path(os.path.dirname(os.getcwd())) / ssh_name
gen_key(ssh_path)
def tunnel():
  !rm /content/text-generation-webui/remomoe
  !ssh -R 80:127.0.0.1:5000 -R 5005:127.0.0.1:5005 -o StrictHostKeyChecking=no -i /content/id_rsa remote.moe > remomoe
threading.Thread(target=tunnel, daemon=True).start()
!sleep 5
!tr -d '\015' < /content/text-generation-webui/remomoe
```

And on the client side, you'd run something like this...

```
ssh -L5005:xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx.remote.moe:5005 remote.moe
```

Once you've done that, you should be good to go, and can just access the API as if it's running locally, usually at `ws://127.0.0.1:5005/api/v1/stream`. For some reason, you can't just use the RemoteMoe URL directly, and I don't know why. Maybe it's possible to use it from a public facing RemoteMoe URL, but I don't know of such a way to do that yet.

**Reply by matthewhand** [reply]
Thanks very much for these instructions!

I had to do a couple of extra steps... download the id_rsa file to my workstation, set perms on it (chmod 600), and pass it as a parameter to the ssh command i.e. -i id_rsa
